src/Gui/Tree.cpp

```diff
@@ -491,16 +500,91 @@ void TreeWidgetItemDelegate::paint(QPainter *painter,
     bool trimColumnSize = isOnlyNameColumnDisplayed();
 
     if (index.column() == 0) {
-        if (tree->testAttribute(Qt::WA_NoSystemBackground)
-                && (trimColumnSize || (opt.backgroundBrush.style() == Qt::NoBrush
-                                && _TreeItemBackground.style() != Qt::NoBrush)))
+        // Only draw custom overlay backgrounds when the tree widget is running
+        // in overlay mode AND the overlay background is translucent. When
+        // overlay is disabled or the background is opaque we must not draw
+        // the extra background (it would clash with the normal widget
+        // rendering). The overlay system sets Qt::WA_NoSystemBackground and
+        // Qt::WA_TranslucentBackground when appropriate, so check both.
+        // Walk up the parent chain to find whether we're inside an overlay
+        // with a translucent background. In some layouts the overlay
+        // attributes/property are set on ancestor widgets (Dock/Overlay
+        // containers) so check parents too. Also consider the "transparent"
+        // property set on OverlayTabWidget.
+    // use file-local helper to detect overlay ancestors
+
+        // Paint the custom overlay background whenever we're inside a
+        // translucent overlay. The background used can be the standard
+        // tree item background (_TreeItemBackground) or, if the user
+        // enabled it, a dedicated overlay color (OverlayActiveColor).
+        // A user-controlled preference can also completely disable the
+        // overlay item background without requiring UI changes. Don't
+        // rely on the style's backgroundBrush (which may change on
+        // hover); only avoid painting over selected or focused items so
+        // the standard selection/active visuals remain visible.
+        // Evaluate painting if either the default tree background is set
+        // or the overlay color feature is enabled.

```

Comment:
I don't see any purpose for this exposition, the code is just a few lines down, and is probably easier to make sense of than this explanation.